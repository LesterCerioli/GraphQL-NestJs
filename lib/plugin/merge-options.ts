import { isString } from '@nestjs/common/utils/shared.utils';

export interface PluginOptions {
  typeFileNameSuffix?: string | string[];
  introspectComments?: boolean;
  advancedTypeIntrospection?: boolean;
}

const defaultOptions: PluginOptions = {
  typeFileNameSuffix: ['.input.ts', '.args.ts', '.entity.ts', '.model.ts'],
  introspectComments: false,
  advancedTypeIntrospection: true,
};

export const mergePluginOptions = (
  options: Record<string, any> = {},
): PluginOptions => {
  if (isString(options.typeFileNameSuffix)) {
    options.typeFileNameSuffix = [options.typeFileNameSuffix];
  }
  return {
    ...defaultOptions,
    ...options,
  };
};
