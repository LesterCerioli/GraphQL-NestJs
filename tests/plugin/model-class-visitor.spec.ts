import * as ts from 'typescript';
import { before } from '../../lib/plugin/compiler-plugin';
import {
  createCatDtoAltText,
  createCatDtoTextAltTranspiled,
} from './fixtures/create-cat-alt.dto';
import {
  createCatDtoText,
  createCatDtoTextTranspiled,
} from './fixtures/create-cat.dto';
import {
  es5CreateCatDtoText,
  es5CreateCatDtoTextTranspiled,
} from './fixtures/es5-class.dto';
import {
  nullableDtoText,
  nullableDtoTextTranspiled,
} from './fixtures/nullable.dto';
import { PluginOptions } from '../../lib/plugin/merge-options';

const defaultCompilerOptions: ts.CompilerOptions = {
  module: ts.ModuleKind.ES2020,
  target: ts.ScriptTarget.ES2020,
  newLine: ts.NewLineKind.LineFeed,
  noEmitHelpers: true,
  strict: true,
};

function transpile(
  source: string,
  pluginOptions: PluginOptions,
  compilerOptions = defaultCompilerOptions,
): string {
  const filename = 'create-cat.input.ts';
  const fakeProgram = ts.createProgram([filename], compilerOptions);

  const result = ts.transpileModule(source, {
    compilerOptions: compilerOptions,
    fileName: 'test.input.ts',
    transformers: {
      before: [before(pluginOptions, fakeProgram)],
    },
  });

  return result.outputText;
}

describe('API model properties', () => {
  it('should add the metadata factory when no decorators exist', () => {
    const actual = transpile(createCatDtoText, { introspectComments: true });
    expect(actual).toMatchSnapshot();
  });

  it('should add partial metadata factory when some decorators exist', () => {
    const options: ts.CompilerOptions = {
      ...defaultCompilerOptions,
      removeComments: true,
    };

    const actual = transpile(
      createCatDtoAltText,
      {
        introspectComments: true,
      },
      options,
    );

    expect(actual).toMatchSnapshot();
  });

  it('should manage imports statements when code "downleveled"', () => {
    const options: ts.CompilerOptions = {
      ...defaultCompilerOptions,
      module: ts.ModuleKind.CommonJS,
      target: ts.ScriptTarget.ES5,
    };

    const actual = transpile(es5CreateCatDtoText, {}, options);
    expect(actual).toMatchSnapshot();
  });

  it('should support & understand nullable type unions', () => {
    const actual = transpile(nullableDtoText, {});
    expect(actual).toMatchSnapshot();
  });

  describe('JsDoc Introspection', () => {
    it('Should parse jsDoc set right before property', () => {
      const source = `
@ObjectType()
export class TestModel {
  /**
  * description
  */
  name: string;
}
      `;
      const actual = transpile(source, { introspectComments: true });

      expect(actual).toMatchInlineSnapshot(`
        "let TestModel = class TestModel {
            static _GRAPHQL_METADATA_FACTORY() {
                return {
                    name: { type: () => String, description: \\"description\\" }
                };
            }
        };
        TestModel = __decorate([
            ObjectType()
        ], TestModel);
        export { TestModel };
        "
      `);
    });

    it('Should parse jsDoc when decorator set on member', async () => {
      const source = `
@ObjectType()
export class TestModel {
  /**
  * description
  */
  @ResolveField(() => Type)
  name: string;
}
`;
      const actual = transpile(source, { introspectComments: true });

      expect(actual).toMatchInlineSnapshot(`
        "let TestModel = class TestModel {
            static _GRAPHQL_METADATA_FACTORY() {
                return {
                    name: { type: () => String, description: \\"description\\" }
                };
            }
        };
        __decorate([
            ResolveField(() => Type)
        ], TestModel.prototype, \\"name\\", void 0);
        TestModel = __decorate([
            ObjectType()
        ], TestModel);
        export { TestModel };
        "
      `);
    });

    it('Should get deprecation reason from @deprecated tag', () => {
      const source = `
@ObjectType()
export class TestModel {
  /**
  * description
  * @deprecated our fancy deprecation message
  */
  @ResolveField(() => Type)
  name: string;
}
`;
      const actual = transpile(source, { introspectComments: true });

      expect(actual).toMatchInlineSnapshot(`
        "let TestModel = class TestModel {
            static _GRAPHQL_METADATA_FACTORY() {
                return {
                    name: { type: () => String, description: \\"description\\", deprecationReason: \\"our fancy deprecation message\\" }
                };
            }
        };
        __decorate([
            ResolveField(() => Type)
        ], TestModel.prototype, \\"name\\", void 0);
        TestModel = __decorate([
            ObjectType()
        ], TestModel);
        export { TestModel };
        "
      `);
    });

    it('Should respect empty @deprecated tag', () => {
      const source = `
@ObjectType()
export class TestModel {
  /**
  * description
  * @deprecated
  */
  @ResolveField(() => Type)
  name: string;
}
`;
      const actual = transpile(source, { introspectComments: true });

      expect(actual).toMatchInlineSnapshot(`
        "let TestModel = class TestModel {
            static _GRAPHQL_METADATA_FACTORY() {
                return {
                    name: { type: () => String, description: \\"description\\", deprecationReason: \\"deprecated\\" }
                };
            }
        };
        __decorate([
            ResolveField(() => Type)
        ], TestModel.prototype, \\"name\\", void 0);
        TestModel = __decorate([
            ObjectType()
        ], TestModel);
        export { TestModel };
        "
      `);
    });

    it('should not parse jsDoc when introspectComments = false', () => {
      const source = `
@ObjectType()
export class TestModel {
  /**
  * description
  * @deprecated
  */
  @ResolveField(() => Type)
  name: string;
}
`;
      const actual = transpile(source, { introspectComments: false });

      expect(actual).toMatchInlineSnapshot(`
        "let TestModel = class TestModel {
            static _GRAPHQL_METADATA_FACTORY() {
                return {
                    name: { type: () => String }
                };
            }
        };
        __decorate([
            ResolveField(() => Type)
        ], TestModel.prototype, \\"name\\", void 0);
        TestModel = __decorate([
            ObjectType()
        ], TestModel);
        export { TestModel };
        "
      `);
    });
  });


  describe('TypeChecker', () => {
    it('Should resolve aliased type', () => {
      const source = `
      
type NumberAlias = number;
type Foo = {bar: 1};

@ObjectType()
export class TestModel {
  // prop: NumberAlias;
  prop2: number[];
  prop2: Foo[];
}
`;
      const actual = transpile(source, { introspectComments: true });

      console.log(actual);
      expect(true).toBeTruthy();
    });

    it('Should resolve optional array enum types in strict mode', () => {
      const source = `
enum Status {
  ENABLED,
  DISABLED
}

@ObjectType()
export class TestModel {
  statusArr?: Status[];
}
`;
      const actual = transpile(source, { introspectComments: true });

      console.log(actual);
      expect(true).toBeTruthy();
    });
  });

  it('Should properly pick parent enum type (not a property) when prop optional in strict mode', () => {
    const source = `
enum Status {
  ENABLED,
  DISABLED
}

@ObjectType()
export class TestModel {
  status2?: Status;
}
`;
    const actual = transpile(source, { introspectComments: true });

    expect(actual).toMatchInlineSnapshot(`
"var Status;
(function (Status) {
    Status[Status[\\"ENABLED\\"] = 0] = \\"ENABLED\\";
    Status[Status[\\"DISABLED\\"] = 1] = \\"DISABLED\\";
})(Status || (Status = {}));
let TestModel = class TestModel {
    static _GRAPHQL_METADATA_FACTORY() {
        return {
            status2: { type: () => Status, nullable: true }
        };
    }
};
TestModel = __decorate([
    ObjectType()
], TestModel);
export { TestModel };
"
`);
  });

  describe('Special virtual types for GraphQL', () => {
    // todo ID / Float / Int
  });


});

// todo:
// repeat test cases for non-type checker implementation
// - separate test-cases for complicated cases:
//   - unions eq 'foo' | 'bar' and string | undefined
//   - promise / observable / Partial

// should throw error when no type checker passed but advancedTypeIntrospection: true
// should work without typeChecker and Program when advancedTypeIntrospection: false
// should skip files with name not matching pattern
