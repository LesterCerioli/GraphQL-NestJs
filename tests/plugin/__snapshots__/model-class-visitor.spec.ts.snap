// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API model properties should add partial metadata factory when some decorators exist 1`] = `
"var Status;
(function (Status) {
    Status[Status[\\"ENABLED\\"] = 0] = \\"ENABLED\\";
    Status[Status[\\"DISABLED\\"] = 1] = \\"DISABLED\\";
})(Status || (Status = {}));
let CreateCatDto2 = class CreateCatDto2 {
    constructor() {
        this.age = 3;
        this.status = Status.ENABLED;
    }
    static _GRAPHQL_METADATA_FACTORY() {
        return {
            name: { type: () => String, description: \\"name description\\" },
            age: { type: () => Number, description: \\"test on age\\" },
            tags: { type: () => [String] },
            status: { type: () => Status },
            breed: { type: () => String, nullable: true },
            nodes: { type: () => [Object] },
            alias: { type: () => Object },
            numberAlias: { type: () => Number },
            union: { type: () => Object },
            intersection: { type: () => Object },
            optionalBoolean: { type: () => Boolean, nullable: true },
            nested: { type: () => Object },
            prop: { type: () => Object },
            tuple: { type: () => Object }
        };
    }
};
CreateCatDto2 = __decorate([
    ObjectType()
], CreateCatDto2);
export { CreateCatDto2 };
"
`;

exports[`API model properties should add the metadata factory when no decorators exist 1`] = `
"var Status;
(function (Status) {
    Status[Status[\\"ENABLED\\"] = 0] = \\"ENABLED\\";
    Status[Status[\\"DISABLED\\"] = 1] = \\"DISABLED\\";
})(Status || (Status = {}));
var OneValueEnum;
(function (OneValueEnum) {
    OneValueEnum[OneValueEnum[\\"ONE\\"] = 0] = \\"ONE\\";
})(OneValueEnum || (OneValueEnum = {}));
let CreateCatDto = class CreateCatDto {
    constructor() {
        this.age = 3;
        this.status = Status.ENABLED;
    }
    static _GRAPHQL_METADATA_FACTORY() {
        return {
            name: { type: () => String, description: \\"description\\" },
            age: { type: () => Number },
            tags: { type: () => [String] },
            status: { type: () => Status },
            status2: { type: () => Status, nullable: true },
            statusArr: { type: () => [Status], nullable: true },
            breed: { type: () => String, nullable: true },
            nodes: { type: () => [Object] },
            date: { type: () => Date },
            booleanProp: { type: () => Boolean },
            stringUnion: { type: () => Object },
            deprecatedField: { type: () => String, description: \\"Testing deprecated tag parsing\\", deprecationReason: \\"this message should go to decorator factory\\" }
        };
    }
};
__decorate([
    HideField()
], CreateCatDto.prototype, \\"hidden\\", void 0);
__decorate([
    ResolveField((type) => Float)
], CreateCatDto, \\"explicitDecorated\\", void 0);
CreateCatDto = __decorate([
    ObjectType()
], CreateCatDto);
export { CreateCatDto };
"
`;

exports[`API model properties should manage imports statements when code "downleveled" 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.CreateCatDtoEs5 = void 0;
var status_1 = require(\\"./status\\");
var constants_1 = require(\\"./constants\\");
var CreateCatDtoEs5 = /** @class */ (function () {
    function CreateCatDtoEs5() {
        this.name = constants_1.CONSTANT_STRING;
        this.status = status_1.ImportedStatus.ENABLED;
        this.obj = constants_1.CONSTANT_OBJECT;
    }
    CreateCatDtoEs5._GRAPHQL_METADATA_FACTORY = function () {
        return {
            name: { type: function () { return String; } },
            status: { type: function () { return Object; } },
            obj: { type: function () { return Object; } }
        };
    };
    CreateCatDtoEs5 = __decorate([
        ObjectType()
    ], CreateCatDtoEs5);
    return CreateCatDtoEs5;
}());
exports.CreateCatDtoEs5 = CreateCatDtoEs5;
"
`;

exports[`API model properties should support & understand nullable type unions 1`] = `
"var Status;
(function (Status) {
    Status[Status[\\"ENABLED\\"] = 0] = \\"ENABLED\\";
    Status[Status[\\"DISABLED\\"] = 1] = \\"DISABLED\\";
})(Status || (Status = {}));
let NullableDto = class NullableDto {
    constructor() {
        this.age = 3;
        this.status = Status.ENABLED;
    }
    static _GRAPHQL_METADATA_FACTORY() {
        return {
            name: { type: () => Object },
            age: { type: () => Number },
            tags: { type: () => [String] },
            status: { type: () => Status },
            status2: { type: () => Object, nullable: true },
            statusArr: { type: () => [Status], nullable: true },
            breed: { type: () => String, nullable: true },
            nodes: { type: () => [Object] },
            date: { type: () => Date }
        };
    }
};
NullableDto = __decorate([
    ObjectType()
], NullableDto);
export { NullableDto };
"
`;
